empress/clumpr/Statistics.py

```diff
@@ -0,0 +1,98 @@
+# of solution costs.
+# https://www.ncbi.nlm.nih.gov/pmc/articles/PMC379178/
+
+from clumpr import DTLReconGraph
+import random
+import matplotlib
+matplotlib.use('tkagg') # need this so plt.show() works
+import matplotlib.pyplot as plt
+
+from typing import List, Dict, Tuple
+
+def trials(host_tree: dict, parasite_tree: dict, phi: dict, dup_cost: float, transfer_cost: float, loss_cost: float, num_trials: int) -> list:
+    """
+    :param host_tree: dictionary representation of host tree
+    :param parasite_tree: dictionary of parasite tree
+    :param phi: dictionary representation of parasite tip to host tip mapping
+    :param dup_cost: float duplication cost
+    :param transfer_cost: float transfer cost
+    :param loss_cost: float loss cost
+    :param num_trials: int number of trials in Monte Carlo simulation
+    :return: list of floating point costs of reconciliations of the Monte Carlo samples
+    """
+   
+    costs = list()  # List of costs of random trials
+    parasites = list(phi.keys())
+    hosts = list(phi.values())
+    for t in range(num_trials):
+        random_phi = create_random_phi(phi)
+        for p in parasites:
+            h = random.choice(hosts)
+            random_phi[p] = h
+        _, cost, _, _ = DTLReconGraph.DP(host_tree, parasite_tree, random_phi, dup_cost, transfer_cost, loss_cost)
+        costs.append(cost)
+    return costs
+
+def create_random_phi(phi : dict) -> dict:
+    """
+    :param phi: dictionary representation of parasite tip to host tip mapping
+    :return: a dictionary of parasite tips to host tips that preserves the degree of
+        of the host tips in phi
+    """
+    random_phi = {}
+    parasites = list(phi.keys())
+    hosts = phi.values()
+    unique_hosts = list(set(hosts))
+    parasite_count = {}   # keys are hosts, values are the number of parasites on that host wrt phi
+    # computer parasite_count values
+    for p in parasites:
+        h = phi[p]
+        if h in parasite_count: parasite_count[h] += 1
+        else: parasite_count[h] = 1
+    for h in unique_hosts:
+        parasite_samples = random.sample(parasites, parasite_count[h])
+        for p in parasite_samples: 
+            random_phi[p] = h
+            parasites.remove(p)
+    return random_phi
+
+def plot_histogram(MPR_cost: int, costs: list):
+    """
+    :param MPR_cost: floating point cost of MPR for given host-parasite-phi and DTL costs
+    :param costs: list of floating point costs of Monte Carlo samples
+    :return: None.  Displays histogram.
+    """
+    plt.hist(costs, color = "blue", density = True)
+    plt.hist(MPR_cost, color="red", rwidth=1)
+    plt.show()
+
+def stats(host_tree: dict, parasite_tree: dict, phi: dict, dup_cost: float, transfer_cost: float, loss_cost: float, num_trials: int) -> (float, list, float):
+    _, MPR_cost, _, _ = DTLReconGraph.DP(host_tree, parasite_tree, phi, dup_cost, transfer_cost, loss_cost)
```

Comment:
This line of code should not go before the comment.

Reply:
I feel like this function should not call `DTLReconGraph.DP` to get the `MPR_cost` (which I think should be renamed to `mpr_cost`). I think the users of this function should call the DP algorithm directly. If we put the DP algorithm here, the DP algorithm might get called multiple times (once for finding the reconciliation tree, and another time for finding the statistics).

Reply:
Moved line before the comment and changed MPR_cost to mpr_cost.  I'm not sure how to resolve the DP issue.  Each trial needs to re-run the DP anew.  It's true that the DP gives us more than what we need here, but we do need to recompute the DP cost each time.  Please feel free to offer another solution - I just don't see it yet. ;^)